### PR TITLE
Rst 7977 add safety areas visualization support to uam driver

### PR DESCRIPTION
**Pull request by carlos-m159**

### CODE REVIEW

**Review by aantunes-locus — APPROVED** [review]

File: include/uam/uam_driver_ros.h [hunk 1]
```diff
@@ -220,6 +249,110 @@ class UamROS
...
+  {
+    auto getLut = [this](const ScanParameters& scan_params, const T& packet)
+    {
+      LaserScanCachedLookup lut;
+      lut.scan_angle_min = scan_params_.getAngleMinLimit();
+      lut.scan_angle_max = scan_params_.getAngleMaxLimit();
+      lut.scan_angle_increment = scan_params_.getAngleIncrement();
+
+      lut.sin_lookup.clear();
+      lut.cos_lookup.clear();
```
Comment (svwilliams) [inline include/uam/uam_driver_ros.h]: Curious why the `clear()` is here? The object was created from scratch a few lines up. Seems like there would never be anything to clear.

File: include/uam/uam_driver_ros.h [hunk 2]
```diff
@@ -220,6 +249,110 @@ class UamROS
...
+
+      lut.sin_lookup.reserve(packet.ranges.size());
+      lut.cos_lookup.reserve(packet.ranges.size());
+
+      for (size_t i = 0; i < packet.ranges.size(); ++i)
```
Comment (svwilliams) [inline include/uam/uam_driver_ros.h:298]: Would there every be a time where `scan_params_.getAngleMaxLimit()` and `scan_params_.getAngleMinLimit() + scan_params_.getAngleIncrement() * (packet.ranges.size() - 1)` would every be different? MaxAngle and (AngleIncrement, Count) are redundant in some sense. Just curious if anything needs to care.

**Review by svwilliams — APPROVED** [review]

File: include/uam/uam_driver_ros.h [hunk 3]
```diff
@@ -190,6 +206,19 @@ class UamROS
       // else is required by the if constexpr
     }
 
+    if (!safety_areas_.empty() && safety_markers_pub_.getNumSubscribers())
+    {
+      if (
+        safety_areas_.count(reply.sensing_data.area_number) &&
+        safety_areas_.at(reply.sensing_data.area_number).count(protocol::EYRAreaType::warning_2))
+      {
+        // Validate incoming ranges agains the warning_2 area
```
Comment (ShivamThukral) [inline include/uam/uam_driver_ros.h]:
```suggestion
        // Validate incoming ranges against the warning_2 area
```

File: src/uam/uam_driver.cpp [hunk 4]
```diff
@@ -246,10 +246,8 @@ std::optional<protocol::YRCommandReply> UamDriver::getSafetyArea(
   const uint32_t end_step)
 {
   // TODO(cribeirmendes): Should we support async send?
-  if (isTheSensorStreaming())
-  {
-    this->stopStreaming();
-  }
+  if (client_.isStopped())
+    client_.startAsyncReadTask();
```
Comment (ShivamThukral) [inline src/uam/uam_driver.cpp:250]:
```suggestion
  if (client_.isStopped())
  {
    client_.startAsyncReadTask();
  }
```

File: include/uam/uam_driver_ros.h [hunk 5]
```diff
@@ -241,6 +371,11 @@ class UamROS
    */
   UamDriver lidar_;
 
+  /**
+   * @brief Lidar safety areas
+   */
+  std::map<size_t, std::map<uam::protocol::EYRAreaType, sensor_msgs::LaserScan>> safety_areas_;
```
Comment (ShivamThukral) [inline include/uam/uam_driver_ros.h]: maybe use `using` and have `std::map<uam::protocol::EYRAreaType, sensor_msgs::LaserScan>` a more meaningful name? Please feel free to ignore this suggestion.

File: include/uam/uam_driver_ros.h [hunk 6]
```diff
@@ -220,6 +249,107 @@ class UamROS
...
+      marker.type = visualization_msgs::Marker::LINE_LIST;
+      marker.ns = ns;
+      marker.scale.x = 0.001;
+      marker.color.a = 1.0;
+      marker.color.r = 1.0;
+      return marker;
+    };
+
+    if (safety_area.ranges.size() != packet.ranges.size())
+      return;
```
Comment (ShivamThukral) [inline include/uam/uam_driver_ros.h:320]:
```suggestion
    if (safety_area.ranges.size() != packet.ranges.size())
    {
      return;
    }
```

**Review by ShivamThukral — APPROVED** [review]

**Pull request merged**